### PR TITLE
feat(union-types): add the support for union types in ruby

### LINKED ISSUE #14: Add support for Union Types
This task is related to adding support for OneOf and AnyOf cases that are supported in OpenApi specifications. More detail can be found [here](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/). The coverage should be complete for the following cases
 - Collection (Array, Map, Array Of Map & Map of Array)
 - Non-Collection (Custom types w/ discriminators, Native types, Enums, Date & DateTime etc)

Also add unit tests to cover all cases.

### PR DESCRIPTION
**Pull request by hamzashoukat94**
## What
This PR adds support for union types that are OneOf and AnyOf types. There are 4 new classes (OneOf, AnyOf, LeafType, UnionTypeContext) that are participating as generic algorithms for validating requests and responses containing union types. You can read more about union types [here](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/). By using those classes, the Ruby core library now has the handling for union types. 

closes #14 
## Why
Add the support for union types in the ruby sdk

## Type of change
Select multiple if applicable.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause a breaking change)
- [ ] Tests (adds or updates tests)
- [ ] Documentation (adds or updates documentation)
- [ ] Refactor (style improvements, performance improvements, code refactoring)
- [ ] Revert (reverts a commit)
- [ ] CI/Build (adds or updates a script, change in external dependencies)

## Dependency Change
If a new dependency is being added, please ensure that it adheres to the following guideline https://github.com/apimatic/apimatic-codegen/wiki/Policy-of-adding-new-dependencies-in-the-core-libraries

## Breaking change
If the PR is introducing a breaking change, please ensure that it adheres to the following guideline https://github.com/apimatic/apimatic-codegen/wiki/Guidelines-for-maintaining-core-libraries

## Testing
List the steps that were taken to test the changes

## Checklist
- [x] My code follows the coding conventions
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added new unit tests

### CODE REVIEW

**Review by sufyankhanrao — COMMENTED** [review]
Blocker: Add newly added classes documentation in the README file.

File: lib/apimatic-core/types/parameter.rb [hunk 1]
```diff
@@ -88,34 +88,24 @@ def get_default_content_type
...
 
-    # Template getter in case a template is set.
-    # @return [Parameter] Returns template for the parameter.
-    def get_template
-      @template
-    end
-
     # Validates the parameter value to be sent in the request.
-    # @raise [ValueError] The value error if the parameter is required but the value is nil.
+    # @raise [ArgumentError] If the parameter is required but the value is nil.
```
Comment (sufyankhanrao) [inline lib/apimatic-core/types/parameter.rb:100]: Nit, Non-blocker: This return docstring is not needed.

File: lib/apimatic-core/utilities/api_helper.rb [hunk 2]
```diff
@@ -228,19 +226,62 @@ def self.clean_url(url)
...
+      # Attempt to deserialize as Boolean
+      return true if value.downcase == 'true'
+      return false if value.downcase == 'false'
+
+      # Default: return the original string
+      value
+    end
+
+    # Checks if a value or all values in a nested structure satisfy a given type condition.
+    # value        - The value or nested structure to be checked.
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/api_helper.rb:259]: Nits, Non-blocker: All param types should be added in this docstring.

File: lib/apimatic-core/utilities/api_helper.rb [hunk 3]
```diff
@@ -228,19 +226,62 @@ def self.clean_url(url)
...
+      else
+        !value.nil? && type_callable.call(value)
+      end
+    end
+
     # Parses JSON string.
     # @param [String] json A JSON string.
-    # rubocop:disable Style/OptionalBooleanParameter
-    def self.json_deserialize(json, should_symbolize = false)
+    def self.json_deserialize(json, should_symbolize = false, apply_primitive_converter = false)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/api_helper.rb:274]: Nits, Non-blocker: `apply_primitive_converter` should be renamed to `allow_primitive_type_parsing`. Also further detail should be added as part of the docstring.

File: lib/apimatic-core/utilities/api_helper.rb [hunk 4]
```diff
@@ -228,19 +226,62 @@ def self.clean_url(url)
...
+    # @return [Object] The deserialized response based on the union_type.
+    def self.deserialize_union_type(union_type, response, should_symbolize = false, should_deserialize = false)
+      response = ApiHelper.json_deserialize(response, false, true) if should_deserialize
+
+      union_type_result = union_type.validate(response)
+
+      union_type_result.deserialize(response, should_symbolize: should_symbolize)
+    end
+
+    def self.apply_primitive_converter(value)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/api_helper.rb:243]: Nits, Non-blocker: `apply_primitive_converter` method name should be renamed to `apply_primitive_type_parser`. Also add method documentation.

File: lib/apimatic-core/utilities/date_time_helper.rb [hunk 5]
```diff
@@ -122,5 +122,52 @@ def self.from_rfc3339(date_time)
         DateTime.rfc3339("#{date_time}Z")
       end
     end
+
+    def self.validate_datetime(dt_format, dt)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/date_time_helper.rb:126]: Nits, Non-blocker: This method named could be renamed to `valid_datetime?`.

File: lib/apimatic-core/utilities/date_time_helper.rb [hunk 6]
```diff
@@ -122,5 +122,52 @@ def self.from_rfc3339(date_time)
...
+      when DateTimeFormat::RFC3339_DATE_TIME
+        return DateTimeHelper.rfc_3339?(dt)
+      when DateTimeFormat::UNIX_DATE_TIME
+        return DateTimeHelper.unix_timestamp?(dt)
+      end
+
+      false
+    end
+
+    def self.validate_date(date_value)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/date_time_helper.rb:139]: Nits, Non-blocker: This method named could be renamed to `valid_date?`.

File: lib/apimatic-core/api_call.rb [hunk 7]
```diff
@@ -86,7 +86,6 @@ def execute
         _deserialized_response = @response_handler.endpoint_logger(@endpoint_logger)
                                                   .endpoint_name_for_logging(@endpoint_name_for_logging)
                                                   .handle(_http_response, @global_configuration.get_global_errors,
-                                                          @global_configuration.get_sdk_module,
                                                           @global_configuration.should_symbolize_hash)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/api_call.rb:89]: Blocker: The `sdk_module` property should also be removed from `GlobalConfiguration` class. As this is not needed anymore.

File: lib/apimatic-core/utilities/union_type_helper.rb [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    end
+
+    def self.invalid_dict_value?(value)
+      value.nil? || !value.instance_of?(Hash)
+    end
+
+    def self.serialize_value(value, context, collection_cases, union_types)
+      return serialize_array_of_dict_case(value, collection_cases) if
+        context.is_array && context.is_dict && context.is_array_of_dict
+      return serialize_dict_of_array_case(value, collection_cases) if
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/union_type_helper.rb:159]: Nits, Non-blocker: Add empty lines after each return to have a better readability.

File: lib/apimatic-core/utilities/union_type_helper.rb [hunk 9]
```diff
@@ -0,0 +1,294 @@
+module CoreLibrary
+  # Helper methods for handling union types.
+  class UnionTypeHelper
+    NONE_MATCHED_ERROR_MESSAGE = 'We could not match any acceptable types against the given JSON.'.freeze
+    MORE_THAN_1_MATCHED_ERROR_MESSAGE = 'There are more than one acceptable type matched against the given JSON.'.freeze
+
+    def self.get_deserialized_value(union_types, value, should_symbolize: false)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/union_type_helper.rb:7]: Nits, Non-blocker: The proximity principle should be applied in this class to keep the closely related methods near to each other.

File: lib/apimatic-core/utilities/union_type_helper.rb [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+      union_types.find(&:is_valid).deserialize(value, should_symbolize: should_symbolize)
+    end
+
+    def self.get_serialized_value(union_types, value)
```
Comment (sufyankhanrao) [inline lib/apimatic-core/utilities/union_type_helper.rb:11]: Nits, Non-blocker: The proximity principle should be applied in this class to keep the closely related methods near to each other.

File: test/test-helper/mock_helper.rb [hunk 11]
```diff
@@ -14,6 +14,8 @@
 require_relative 'models/attributes_and_elements'
 require_relative 'models/atom'
 require_relative 'models/morning'
+require_relative 'models/month_name_enum'
+require_relative 'models/month_number_enum'
```
Comment (sufyankhanrao) [inline test/test-helper/mock_helper.rb:18]: Questions: Are these imports needed?

File: test/test-helper/models/scalar_model.rb [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+      # Create object from extracted values.
+      ScalarModel.new(any_of_required,
+                      one_of_req_nullable,
+                      one_of_optional,
+                      any_of_opt_nullable)
+    end
+
+    # Validates an instance of the object from a given value.
+    # @param [ScalarModel | Hash] The value against the validation is performed.
+    def self.validate(dictionary)
```
Comment (sufyankhanrao) [inline test/test-helper/models/scalar_model.rb:90]: Blocker: The newly generated models contain the changes in validation whenever the model property is OneOf/AnyOf. I'd recommend to regenerate this model from the code-gen branch for SDKs.

File: test/test-helper/models/month_name_enum.rb [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+      OCTOBER = 'October'.freeze,
+
+      # TODO: Write general description for NOVEMBER
+      NOVEMBER = 'November'.freeze,
+
+      # TODO: Write general description for DECEMBER
+      DECEMBER = 'December'.freeze
+    ].freeze
+
+    def validate(value)
```
Comment (sufyankhanrao) [inline test/test-helper/models/month_name_enum.rb:42]: Blocker: The newly generated enums through the code-gen SDK branch contain changes for this validate method. I'd recommend to update the enum classes with newly generated classes.

File: test/test-apimatic-core/response_handler_test.rb [hunk 14]
```diff
@@ -271,6 +272,25 @@ def test_primitive_response_body
...
+    response_mock = MockHelper.create_response status_code: 200,
+                                               raw_body: response_body_mock
+    actual_response = @response_handler.deserializer(ApiHelper.method(:deserialize_primitive_types))
+                                       .is_primitive_response(true)
+                                       .deserializer(proc do |response, should_symbolize|
+                                         ApiHelper.deserialize_union_type(
+                                           AnyOf.new([LeafType.new(Float, UnionTypeContext.new(is_array: true)),
+                                                      LeafType.new(String)]), response, should_symbolize, true)
+                                       end)
+                                       .handle(response_mock, MockHelper.get_global_errors, TestComponent)
```
Comment (sufyankhanrao) [inline test/test-apimatic-core/response_handler_test.rb:286]: Blocker: The `handle` method does not expect sdk_module anymore therefore the TestComponent as last param is not required now. This should be removed from all test cases where applicable.

File: test/test-apimatic-core/utilities/api_helper_test.rb [hunk 15]
```diff
@@ -339,27 +339,6 @@ def test_complex_type_query_params
...
-    assert obj.instance_of?(TestComponent::Atom)
-    obj = ApiHelper.deserialize('oneOf(Morning, Atom, Car)',
-                          '{"startsAt":"6:00","endsAt":"11:00","offerTeaBreak":true,"sessionType":'\
-      '"Morning"}', TestComponent, false
-    )
-    assert obj.instance_of?(TestComponent::Morning)
-
-  end
-
   def test_json_serialize
```
Comment (sufyankhanrao) [inline test/test-apimatic-core/utilities/api_helper_test.rb:342]: Blocker: The unit tests for the newly added methods like `valid_type`, `apply_primitive_converter`, `deserialize_union_type` are missing.

File: test/test-apimatic-core/union_type_tests/one_of_test.rb [hunk 16]
```diff
@@ -0,0 +1,1745 @@
+require 'minitest/autorun'
+require 'apimatic_core'
+
+require_relative '../../test-helper/models/morning'
+require_relative '../../test-helper/models/evening'
+require_relative '../../test-helper/models/month_number_enum'
+require_relative '../../test-helper/models/month_name_enum'
+
+class OneOfTest < Minitest::Test
```
Comment (sufyankhanrao) [inline test/test-apimatic-core/union_type_tests/one_of_test.rb:9]: Question: Have we added unit tests for `nested OneOf cases`? If not, kindly add at-least 1 test case for that.

File: test/test-apimatic-core/union_type_tests/one_of_test.rb [hunk 17]
```diff
@@ -0,0 +1,1745 @@
...
+
+  def test_valid_mutiple_types_in_one_of
+    _one_of = OneOf.new([LeafType.new(Integer), LeafType.new(Float), LeafType.new(String)])
+    _one_of.validate(4)
+    assert _one_of.is_valid
+  end
+
+  def test_valid_both_string_one_of
+    _one_of = OneOf.new([LeafType.new(String), LeafType.new(String)])
+    assert_raises OneOfValidationException do
```
Comment (sufyankhanrao) [inline test/test-apimatic-core/union_type_tests/one_of_test.rb:58]: Blocker: The assertion against the error messages that the exception has thrown is missing. I'd recommend to see the tests in python core library.

File: test/test-apimatic-core/union_type_tests/any_of_test.rb [hunk 18]
```diff
@@ -0,0 +1,2461 @@
...
+  end
+
+  def test_invalid_any_of
+    _any_of = AnyOf.new(
+      [
+        LeafType.new(String),
+        LeafType.new(String)
+      ]
+    )
+    assert_raises AnyOfValidationException do
```
Comment (sufyankhanrao) [inline test/test-apimatic-core/union_type_tests/any_of_test.rb:85]: Blocker: The assertion against the error messages that the exception has thrown is missing. I'd recommend to see the tests in python core library.

File: test/test-apimatic-core/union_type_tests/one_of_any_of_test.rb [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+          [
+            LeafType.new(String),
+            LeafType.new(String)
+          ]
+        ),
+        LeafType.new(TrueClass),
+        LeafType.new(FalseClass)
+      ]
+    )
+    assert_raises AnyOfValidationException do
```
Comment (sufyankhanrao) [inline test/test-apimatic-core/union_type_tests/one_of_any_of_test.rb:48]: Blocker: The assertion against the error messages that the exception has thrown is missing. I'd recommend to see the tests in python core library.

**Review by sufyankhanrao — CHANGES_REQUESTED** [review]
Above mentioned changes need your attention.

File: test/test-helper/mock_helper.rb (hunk 11 above)
Comment (hamzashoukat94) [reply test/test-helper/mock_helper.rb:18]: Yes these imports are needed

File: lib/apimatic-core/types/parameter.rb (hunk 1 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/types/parameter.rb:100]: Done

File: lib/apimatic-core/utilities/api_helper.rb (hunk 2 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/api_helper.rb]: added

File: lib/apimatic-core/utilities/api_helper.rb (hunk 3 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/api_helper.rb]: done

File: lib/apimatic-core/utilities/api_helper.rb (hunk 4 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/api_helper.rb]: done

File: lib/apimatic-core/utilities/date_time_helper.rb (hunk 5 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/date_time_helper.rb]: done

File: lib/apimatic-core/utilities/date_time_helper.rb (hunk 6 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/date_time_helper.rb]: done

File: lib/apimatic-core/api_call.rb (hunk 7 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/api_call.rb:89]: done

File: lib/apimatic-core/utilities/union_type_helper.rb (hunk 8 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/union_type_helper.rb:152]: done

File: lib/apimatic-core/utilities/union_type_helper.rb (hunk 9 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/union_type_helper.rb]: done

File: lib/apimatic-core/utilities/union_type_helper.rb (hunk 10 above)
Comment (hamzashoukat94) [reply lib/apimatic-core/utilities/union_type_helper.rb]: done

File: test/test-apimatic-core/response_handler_test.rb (hunk 14 above)
Comment (hamzashoukat94) [reply test/test-apimatic-core/response_handler_test.rb]: done

File: test/test-apimatic-core/union_type_tests/one_of_test.rb (hunk 17 above)
Comment (hamzashoukat94) [reply test/test-apimatic-core/union_type_tests/one_of_test.rb]: done

File: test/test-apimatic-core/union_type_tests/any_of_test.rb (hunk 18 above)
Comment (hamzashoukat94) [reply test/test-apimatic-core/union_type_tests/any_of_test.rb]: done

File: test/test-helper/models/month_name_enum.rb (hunk 13 above)
Comment (hamzashoukat94) [reply test/test-helper/models/month_name_enum.rb:42]: done

File: test/test-helper/models/scalar_model.rb (hunk 12 above)
Comment (hamzashoukat94) [reply test/test-helper/models/scalar_model.rb:90]: done

File: test/test-apimatic-core/utilities/api_helper_test.rb (hunk 15 above)
Comment (hamzashoukat94) [reply test/test-apimatic-core/utilities/api_helper_test.rb:342]: done

**Review by sufyankhanrao — DISMISSED** [review]
LGTM

**Review by sufyankhanrao — APPROVED** [review]

**Review by sufyankhanrao — APPROVED** [review]

**Pull request merged**